npm-publish/bin/publish.sh

```diff
@@ -0,0 +1,137 @@
+#!/usr/bin/env bash
+
+set -o errexit   # exit on error
+set -o errtrace  # exit on error within function/sub-shell
+set -o nounset   # error on undefined vars
+set -o pipefail  # error if piped command fails
+
+# Default variables
+RELEASE_TYPE=
+MAIN_BRANCH="trunk"
+
+echo_title() {
+	echo ""
+	echo "== $1 =="
+}
+
+while getopts ":t:" option;
+do
+	case $option in
+		# npm major/minor/patch
+		t) RELEASE_TYPE=$OPTARG ;;
+
+		\?) echo "Error: Invalid param / option specified"
+			exit 199 ;;
+   esac
+done
+
+# Validate release type value
+if [ "$RELEASE_TYPE" != "major" ] && [ "$RELEASE_TYPE" != "minor" ] && [ "$RELEASE_TYPE" != "patch" ]; then
+	echo "❌ Invalid release type specified. Please make sure the -t flag is one of major/minor/patch."
+	exit 200
+fi
+
+# Fetch some basic package information
+echo_title "Fetching local package info"
+LOCAL_NAME=$(node -p "require('./package.json').name")
+LOCAL_VERSION=$(node -p "require('./package.json').version")
+LOCAL_BRANCH=$(git branch --show-current)
+REMOTE_VERSION=$(npm view "$LOCAL_NAME" version)
+echo "✅ Found $LOCAL_NAME $LOCAL_VERSION on branch $LOCAL_BRANCH"
+echo "✅ Published version is $REMOTE_VERSION"
+echo "✅ Will publish new $RELEASE_TYPE release"
+
+# Validate npm is logged in and ready
+echo_title "Checking npm auth"
+if ! NPM_USER=$( npm whoami ); then
+	echo "❌ npm cli is not authenticated. Please make sure you're logged in or NPM_TOKEN is set."
+	exit 201
+fi
+echo "✅ Logged in as $NPM_USER and ready to publish"
+
+# Validate current branch
+echo_title "Checking branch"
+# TODO: add support for release/** branch via [ ! "$LOCAL_BRANCH" == "$RELEASE_BRANCH_PATTERN" ] -- will need to add version checking support as well
+if [ "$LOCAL_BRANCH" != "$MAIN_BRANCH" ]; then
+	echo "❌ You can only publish from the '$MAIN_BRANCH' branch. Please switch branches and try again."
+	exit 202
+fi
+echo "✅ On a valid release branch ($LOCAL_BRANCH)"
+
+# Validate no uncommitted changes.
+# Shouldn't happen in CI but protects against local runs.
+echo_title "Checking for local changes"
+if ! git diff-index --quiet HEAD --; then
+	echo "❌ Working directory has uncommitted changes; please clean up before proceeding."
+	exit 203
+fi
+echo "✅ No local changes found"
+
+### TODO=====================
+# Validate version not published
+# Need a cleaner way to fetch published versions
+#echo_title "Checking version"
+#IS_VERSION_PUBLISHED=$(npm info . versions --json | grep -q "\"$LOCAL_VERSION\"")
+#echo $IS_VERSION_PUBLISHED
+### TODO=====================
+
+# Install
+echo_title "npm ci"
+npm ci
+echo "✅ Clean install looks good"
+
+# Test
+echo_title "npm test"
+npm test
+echo "✅ Tests look good"
```

Comment:
We can possibly combine install and test steps: [`npm cit`](https://docs.npmjs.com/cli/v6/commands/npm-install-ci-test) = `npm ci` + `npm test`

Reply:
Or, from the view of security, we could do something like

```bash
npm ci --ignore-scripts
NODE_AUTH_TOKEN= npm rebuild
NODE_AUTH_TOKEN= npm run prepare --if-present
NODE_AUTH_TOKEN= npm test
```

Then, if we have a malicious package (bad things happen - remember the story with the `colors` package :-( ), it won't be able to steal our NPM authentication token.

We will use the token to install the dependencies (and tell `npm` not to run any pre/postinstall scripts), then we run `npm rebuild`, `npm run prepare --if-present`, and `npm test` without the `NODE_AUTH_TOKEN`. Thus, if a malicious package tries to scan the environment for the auth token, it won't be able to steal it.